[quant] Quantizable MultiheadAttention

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51169 [quant][refactor] Factor out MHA code dup from nn and nn.quantizable
* #50459 [quant] Factoring out the list of no_observers
* **#49866 [quant] Quantizable MultiheadAttention**

- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

Commit:
[quant][WIP] Quantizable MultiheadAttention

This is still work in progress, and the tests will be amended into this PR.

[ghstack-poisoned]

Commit:
Update on "[quant][WIP] Quantizable MultiheadAttention"


This is still work in progress, and the tests will be amended into this PR.

[ghstack-poisoned]

Commit:
Update on "[quant][WIP] Quantizable MultiheadAttention"


This is still work in progress, and the tests will be amended into this PR.

[ghstack-poisoned]

Commit:
Update on "[quant][WIP] Quantizable MultiheadAttention"


This is still work in progress, and the tests will be amended into this PR.

[ghstack-poisoned]

Commit:
Update on "[quant][WIP] Quantizable MultiheadAttention"


This is still work in progress, and the tests will be amended into this PR.

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant][WIP] Quantizable MultiheadAttention"


This is still work in progress, and the tests will be amended into this PR.

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant][WIP] Quantizable MultiheadAttention"


This is still work in progress, and the tests will be amended into this PR.

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant][WIP] Quantizable MultiheadAttention"


This is still work in progress, and the tests will be amended into this PR.

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant][WIP] Quantizable MultiheadAttention"


This is still work in progress, and the tests will be amended into this PR.

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant][WIP] Quantizable MultiheadAttention"


This is still work in progress, and the tests will be amended into this PR.

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant][WIP] Quantizable MultiheadAttention"


This is still work in progress, and the tests will be amended into this PR.

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant][WIP] Quantizable MultiheadAttention"


This is still work in progress, and the tests will be amended into this PR.

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

vkuzo commented on torch/nn/quantizable/modules/activation.py:
```diff
@@ -0,0 +1,344 @@
+class MultiheadAttention(nn.MultiheadAttention):
+    _FLOAT_MODULE = nn.MultiheadAttention
+
+    r"""Allows the model to jointly attend to information
```
optional: could we refer to the original docs without copy-paste?  ex: https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/activation.py#L91

vkuzo replied:
IMO it would be good for this docblock to explain why this class exists compared to the original MHA class

vkuzo commented on torch/nn/quantizable/modules/activation.py:
```diff
@@ -0,0 +1,344 @@
+    def forward(self, query, key, value, key_padding_mask=None,
+                need_weights=True, attn_mask=None):
+        # type: (Tensor, Tensor, Tensor, Optional[Tensor], bool, Optional[Tensor]) -> Tuple[Tensor, Optional[Tensor]]
+        r"""
```
same here, assuming that this docblock is the same as the fp32 module

vkuzo commented on torch/quantization/quantize.py:
```diff
@@ -155,7 +155,7 @@ def insert_activation_post_process(m, special_act_post_process=None):
             # TODO: These are the modules that cannot be observed
             #       Once there are more, we should move them to a separate list
-            if custom_module_class_mapping[type(child)] != nnqa.LSTM:
+            if custom_module_class_mapping[type(child)] not in [nnqa.LSTM, nnqa.MultiheadAttention]:
```
should we move this into a list in the config?

vkuzo replied:
sounds great, would it be ok to do this in this PR?

z-a-f replied:
After thinking a moment -- I will send another PR with the mappings stacked on top of this. I think it's a bad idea to piggyback new features with the existing stuff because in case of a bisect, the whole thing might go down.

vkuzo replied:
sg

vkuzo replied:
> I will send another PR with the mappings stacked on top of this

I'm ok with not doing it here, but for the future would be great to consider stacking the one-off fixes before the large logic change PR (instead of after), to reduce the amount of time the hacks live in the master branch to zero.

vkuzo commented on torch/nn/quantizable/modules/activation.py:
```diff
@@ -0,0 +1,344 @@
+        return self._forward_impl(query, key, value, key_padding_mask,
+                                  need_weights, attn_mask)
+
+    def _forward_impl(
```
high level, which code here is copy-pasta, and which is specific to this file?  Do we need to do some refactoring to reuse more code from the original version?

vkuzo replied:
>  The problem is that because we have to be stateful for quantization to work, I cannot use the code from there directly, unless I refactor the torch.nn.functional.mult_head_attention (I really wanna do that tho -- would require me changing the nn namespace :(

Yeah, that's exactly my question.  I understand if in the end this is the best way to go, I'm asking if we have considered the tradeoffs, reached out to the people who wrote original MHA, etc.  Copy pasting this is a non-trivial amount of tech debt, so just wanted to get more context on how we ended up on the current approach.

z-a-f replied:
The functional call to the MHA (either in forward or as a separate function) is always gonna be the same. The problem is factoring out code from the `nn.functional` for it to be used in the `nn.quantized` might be mildly tricky -- will take a look

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

z-a-f commented:
@vkuzo I added a reference to the FP docstring, but I still believe that the docstrings here should have descriptions about the args, even if they are copied. That way the user doesn't need to jump around the documentation to get info about the input shapes. Let me know if you still believe we should drop the args from here, and just refer to the original -- I'll do it

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

vkuzo commented on torch/nn/quantizable/modules/activation.py:
```diff
@@ -0,0 +1,356 @@
+
+        if self.add_zero_attn:
+            src_len += 1
+            k_zeros = torch.zeros((k.size(0), 1) + k.size()[2:])
+            if k.is_quantized:
+                k_zeros = torch.quantize_per_tensor(k_zeros, k.q_scale(), k.q_zero_point(), k.dtype)
```
would there be a way to rewrite these with `QuantStub` instead of using `quantize_per_tensor` directly?

z-a-f replied:
That is not necessary at all -- the k_zeros is always a tensor of zeroes, and the quant_stub will try to observe it, which is not necessary. It is better to quantize it with the parameters of `k` due to concatenation later.

vkuzo commented on torch/nn/quantizable/modules/activation.py:
```diff
@@ -0,0 +1,356 @@
+
+        q = self.q_scaling_product.mul_scalar(q, scaling)
+
+        if attn_mask is not None:
```
I might be missing something, but seems like L241 to L299 are a direct copy? Can we move this to a util function and reuse instead?

z-a-f replied:
Do you mean factor out code from the `nn.functional`?

raghuramank100 replied:
Agree, this will require refactoring the nn.functional MHA code, but it is worth it to avoid duplication.

z-a-f replied:
All the refactoring will be part of the upcoming PRs

vkuzo replied:
sg

vkuzo commented on torch/nn/quantizable/modules/activation.py:
```diff
@@ -0,0 +1,356 @@
+        attn_output_weights = torch.bmm(q, k.transpose(1, 2))
+        assert list(attn_output_weights.size()) == [bsz * self.num_heads, tgt_len, src_len]
+
+        if attn_mask is not None:
+            if attn_mask.dtype == torch.bool:
+                attn_output_weights.masked_fill_(attn_mask, float('-inf'))
+            else:
+                attn_output_weights += attn_mask
+
+        if key_padding_mask is not None:
+            attn_output_weights = attn_output_weights.view(bsz, self.num_heads, tgt_len, src_len)
+            attn_output_weights = attn_output_weights.masked_fill(
+                key_padding_mask.unsqueeze(1).unsqueeze(2),
+                float('-inf'),
+            )
+            attn_output_weights = attn_output_weights.view(bsz * self.num_heads, tgt_len, src_len)
+
+        attn_output_weights = nnF.softmax(
+            attn_output_weights, dim=-1)
+        attn_output_weights = nnF.dropout(attn_output_weights, p=self.dropout, training=self.training)
+
+        attn_output = torch.bmm(attn_output_weights, v)
+        assert list(attn_output.size()) == [bsz * self.num_heads, tgt_len, head_dim]
+        attn_output = attn_output.transpose(0, 1).contiguous().view(tgt_len, bsz, self.embed_dim)
```
seems like this is a direct copy as well?

z-a-f replied:
All the refactoring will be in the upcoming PRs

z-a-f commented:
@vkuzo I was trying to factor out the stuff in the `nn.functional` to reduce the dupilicate code -- it is not as straightforward as I initially thought. I believe it is better to do it as a separate PR, so that bisect would be able to find if it causes any issues. I am adding a comment in the forward method for future refactoring

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

vkuzo commented:
> @vkuzo I was trying to factor out the stuff in the `nn.functional` to reduce the dupilicate code -- it is not as straightforward as I initially thought. I believe it is better to do it as a separate PR, so that bisect would be able to find if it causes any issues. I am adding a comment in the forward method for future refactoring

makes sense, to clarify, does this mean separate PR in the same stack landed at the same time, or sometime in the future?

z-a-f commented:
> > @vkuzo I was trying to factor out the stuff in the `nn.functional` to reduce the dupilicate code -- it is not as straightforward as I initially thought. I believe it is better to do it as a separate PR, so that bisect would be able to find if it causes any issues. I am adding a comment in the forward method for future refactoring
> 
> makes sense, to clarify, does this mean separate PR in the same stack landed at the same time, or sometime in the future?

Might send it later on -- a little busy at the moment

raghuramank100 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2395,6 +2395,100 @@ def test_custom_module_lstm(self):
+        batch_size = 4
+        target_seq_length = 128
+        source_seq_length = 64
+        embed_dim = 512  # Must be divisible buy the number of heads
```
Typo "buy"

raghuramank100 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2395,6 +2395,100 @@ def test_custom_module_lstm(self):
+                    Bias, Add_bias_kv, Add_zero_attn):
+                # Assume 12dB is sufficient for functional equivalence
+                # Without the bias, linear might perform poorly
+                min_power = 12 if bias else 10
```
Can we do a tighter check? If you can use the dequantized versions of the quantized MHA weights, then the outputs should match very closely.

z-a-f replied:
In my tests the snr was pretty low -- I'll try the q/dq weights

raghuramank100 commented on torch/nn/quantizable/modules/activation.py:
```diff
@@ -0,0 +1,368 @@
+    def from_float(cls, other):
+        assert type(other) == cls._FLOAT_MODULE
+        assert hasattr(other, 'qconfig'), "The float module must have 'qconfig'"
+        # Setting the dropout to 0.0!
```
Can we support QAT with this module?  We should not set the dropout to zero here, this will be handled when the model is set to eval.

z-a-f replied:
The `prepare_qat` is not implemented, so there is not need for this PR

raghuramank100 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2395,6 +2395,100 @@ def test_custom_module_lstm(self):
                              f"Signal: {signal}, MSE: {mse}"))
 
+    @override_qengines
+    def test_custom_module_multi_head_attention(self):
```
Can we also add a test for QAT, i.e ensuring that we can create a model with fake quant after prepare_qat, followed by convert?

z-a-f replied:
Prepare_qat is not implemented -- will send it as a separate PR

raghuramank100 commented on torch/nn/quantizable/modules/activation.py:
```diff
@@ -0,0 +1,368 @@
+                                               inplace=False,
+                                               remove_qconfig=True,
+                                               convert_custom_config_dict=None)
+        # Remove the parameters for the bias_k and bias_v to quantize them
```
We do not need to quantize the bias, our kernels handle them in fp32

z-a-f replied:
bias_k is not a bias for linear -- it's additional column in the key tensor. Same for the bias_v

raghuramank100 commented on torch/nn/quantizable/modules/activation.py:
```diff
@@ -0,0 +1,368 @@
+            k_zeros = torch.zeros((k.size(0), 1) + k.size()[2:])
+            if k.is_quantized:
+                k_zeros = torch.quantize_per_tensor(k_zeros, k.q_scale(), k.q_zero_point(), k.dtype)
+            k = torch.cat([k, k_zeros], dim=1)
```
What is the scale and zero-point of k at the output? Does the functional cat use the same output scale for k?

z-a-f replied:
The functional cat uses the scale and zp of the first element in the list

vkuzo reviewed (changes requested):
chatted offline, requesting changes for either reusing code instead of copy-pasting, or getting some alignment from PyTorch core that copy-pasta is the right way to go.  If we do decide to copy, can we document the reasoning in the PR summary and the code?

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

vkuzo commented on torch/nn/quantizable/modules/activation.py:
```diff
@@ -0,0 +1,439 @@
+import torch
```
nit on the filename, is `activation.py` the best filename for it?  I would have expected something like `multiheadattention.py`, etc

z-a-f replied:
MHA is a type of activation by convention -- it's more of a head (imho), but pytorch does not have a "head" subgroup for the layers, so it is put in the activation group

vkuzo commented on torch/nn/quantizable/modules/activation.py:
```diff
@@ -0,0 +1,439 @@
+        self.dequant_v = torch.quantization.DeQuantStub()
+
+    def _get_name(self):
+        return 'QuantizableMultiheadAttention'
```
is this going to return `QuantizableMultiheadAttention` in both fp32 with observers and quantized state?  If yes, would it be possible to refactor the code so that the final quantized module prints out `QuantizedMultiheadAttention`, to help with debugging?

z-a-f replied:
It is not as straight-forward. This will require reflection design pattern with the change of function when the `from_observed` is called. I would not recommend doing it, as this is a power-feature. Alternatively, one can check if the layer weights are quantized, and change the name accordingly. But I think it might be error prone -- I'll think about it...

vkuzo replied:
it would be good to fix this in the long term, the current state will be confusing to people who use this without reading the code

vkuzo reviewed (approved):
lg, thanks for addressing the comments!

raghuramank100 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2401,6 +2401,102 @@ def test_custom_module_lstm(self):
+            for bias, add_bias_kv, add_zero_attn in itertools.product(
+                    Bias, Add_bias_kv, Add_zero_attn):
+                min_power = 20
+                max_mse = 4
```
Can we have a tighter check on the quantized module vs expected?  Perhaps having a reference op that does exactly the same numerics with fp32 operations can be done.

z-a-f replied:
I already implemented the `dequantize` method inside the module that creates a dequantized reference -- the numerics don't exceed 20dB

mehrdad78 replied:
hi z-a-f
now,how can i dynamic_quantize wav2vec model?

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]

Commit:
Update on "[quant] Quantizable MultiheadAttention"


- Adds the `torch.nn.quantizable.MultiheadAttention`

The quantizable version can serve as a fully equivalent to `torch.nn.MultiheadAttention` module.
The main difference is that it allows for linear units observation after the `prepare` step in the quantization flow.

Note: The `from_observed` method (called during the `convert`) removes the `bias_k` and `bias_v` parameters, and resets them as attributes.
This is done to avoid an error of assigning a quantized tensor to the `torch.nn.Parameter`.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_multi_head_attention
```

Differential Revision: [D25706179](https://our.internmc.facebook.com/intern/diff/D25706179)

[ghstack-poisoned]
